### PR TITLE
Migrate torch::deploy test infra to pytorch/test-infra

### PR DESCRIPTION
**Pull request by hatala91**
With this PR, I propose to migrate [runtime_nightly.yaml](https://github.com/pytorch/multipy/blob/main/.github/workflows/runtime_nightly.yaml) and [runtime_tests.yaml](https://github.com/pytorch/multipy/blob/main/.github/workflows/runtime_tests.yaml) to use [pytorch/test_infra](https://github.com/pytorch/test-infra/wiki/).

For this, I introduce a new reusable workflow in [build_test_release.yaml](https://github.com/pytorch/multipy/pull/274/files#diff-e9376faa19019165b5c9299c6e79ef11820184ea067877dd574bf357b7a7e01b) which is called from both the runtime tests and nightly build pipeline. The workflow then uses `pytorch/test_infra's` [linux_job.yml](https://github.com/pytorch/test-infra/blob/main/.github/workflows/) to build, test, run examples and benchmark.

Example runs:
 - py37: https://github.com/pytorch/multipy/actions/runs/3708138595/jobs/6285368845
 - py38: https://github.com/pytorch/multipy/actions/runs/3708138595/jobs/6285368958
 - py39: https://github.com/pytorch/multipy/actions/runs/3708138595/jobs/6285369062
 - py310: https://github.com/pytorch/multipy/actions/runs/3708138595/jobs/6285368633

Issue: https://github.com/pytorch/multipy/issues/254

### CODE REVIEW

File: .github/workflows/build_test_release.yaml
Comment (PaliC) [inline .github/workflows/build_test_release.yaml:36]: Is it possible to split these up into different jobs? You should be able to create a dependency from the build and the other jobs. Just make sure tarball creation + distribution is dependent on the tests passing.

File: .github/workflows/build_test_release.yaml
Comment (PaliC) [inline .github/workflows/build_test_release.yaml:66]: Just add a comment for why you are doing the renaming as its not intuitive and relies on an implementation detail. I'll bug the code owner (@seemethere) to see if I can add tar support to simplify this.

**Review by PaliC — CHANGES_REQUESTED** [review]

File: .github/workflows/build_test_release.yaml
Comment (hatala91) [reply .github/workflows/build_test_release.yaml]: Yes, that would be possible! However, each step using the `linux-job` from `pytest/test-infra` begins with a plain [conda-builder](https://hub.docker.com/r/pytorch/conda-builder) image. Therefore, each would need to install dependencies and build MultiPy separately which is quite expensive I'm afraid.

I added what I believe to be the next best thing which is log groups (sections that can be collapsed):

image

File: .github/workflows/build_test_release.yaml
Comment (hatala91) [reply .github/workflows/build_test_release.yaml]: Yes, great idea! I added a comment with a todo to remove the renaming once https://github.com/pytorch/test-infra/pull/1262 is merged.

File: .github/workflows/build_test_release.yaml
Comment (PaliC) [reply .github/workflows/build_test_release.yaml]: ohh that's great. We just want to make sure we have a good way to split things up :)

File: multipy/runtime/loader.cpp
Comment (PaliC) [inline multipy/runtime/loader.cpp:70]: Is this for CentoOS 7? Can you a bit more clear on why its needed.

File: .github/workflows/build_test_release.yaml
Comment (PaliC) [inline .github/workflows/build_test_release.yaml:122]: @seemethere had some comments on your test-infra PR as a way to improve this. Happy to approve after you fix that.

**Review by PaliC — CHANGES_REQUESTED** [review]
This is almost there. You just need to address the distribution bit.

Overall great work! This is a major improvement to our test infra in terms of better engineering :)

File: .github/workflows/build_test_release.yaml
Comment (hatala91) [reply .github/workflows/build_test_release.yaml:122]: Adopted (even pushed that commit shortly before you commented) and can confirm that moving the artifact directly to that folder works excellent:

image

File: .github/workflows/build_test_release.yaml
Comment (hatala91) [reply .github/workflows/build_test_release.yaml]: 🦅👀

File: multipy/runtime/loader.cpp
Comment (hatala91) [reply multipy/runtime/loader.cpp]: Great catch, I copied the whole block and forgot to adapt the comment.

**Review by PaliC — APPROVED** [review]
Nice! This is amazing. 

As some logistical asides before landing. Please add a summary to describe your changes. (I think the issue should give decent motivation, but its generally good practice to add a few sentences or more describing what you did.)

Also add some jobs to the summary to show that release and build-test work as intended.

**Review by seemethere — APPROVED** [review]
This is awesome!

**Pull request closed without merging**
